### PR TITLE
feature/router_macro

### PR DESCRIPTION
**Pull request by boyswan**
This PR adds the `path!` router proc_macro.

As well as constructing an n-tuple of `StaticSegment | ParamSegment | StaticSegment`, the macro errors on a handful of invalid cases:

- deny consecutive_slashes  
example: `//`

- deny non_trailing_wildcard 
example:`/home/*any/end`

- deny invalid segment 
example: `/foo/^/"`
characters that are not `ascii_lowercase`, `asci_digit` or `-` are deemed invalid

### CODE REVIEW

File: router_macro/src/lib.rs [hunk 1]
```diff
@@ -45,32 +41,102 @@ impl SegmentParser {
...
+            }
+        }
+    }
+}
+
+impl Segment {
+    fn is_valid(segment: &str) -> bool {
+        segment
+            .chars()
+            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-')
```
Comment (gbj) [inline router_macro/src/lib.rs:85]: I'm wondering if this is slightly too restrictive — for example, I'd think `_` should be a valid character for static segments, at least, and possibly other [rarer but spec-compliant characters](https://stackoverflow.com/questions/4669692/valid-characters-for-directory-part-of-a-url-for-short-links).

**Review by gbj — COMMENTED** [review]
Thanks, this looks great overall! See one comment about permitted characters.

File: router_macro/src/lib.rs (hunk 1 above)
Comment (boyswan) [reply router_macro/src/lib.rs]: Great point!

I've added the following permitted characters:
```rust
const RFC3986_UNRESERVED: [char; 4] = ['-', '.', '_', '~'];
const RFC3986_PCHAR_OTHER: [char; 1] = ['@'];
```

And also added uppercase ascii:
```rust
c.is_ascii_digit()
    || c.is_ascii_lowercase()
    || c.is_ascii_uppercase()
    || RFC3986_UNRESERVED.contains(&c)
    || RFC3986_PCHAR_OTHER.contains(&c)
```

Technically `RFC3986_PCHAR_OTHER` also permits `:`, however I've removed it due to potential clashes with param logic.

Comment (gbj) [reply router_macro/src/lib.rs]: Thanks! This seems good to me. If someone really `:` in a static route they can always use the tuple syntax... This is just a convenience.

**Pull request merged**